File: without-library/graph/components/bipartite_check.py
import sys

# Bipartite Check (BFS)

class Bipartite:
    def __init__(self, n, adj):
        self.n = n
        self.adj = adj
        self.color = [-1] * (n + 1)
        
    def check(self, start):
        queue = [start]
        head = 0
        self.color[start] = 0
        
        while head < len(queue):
            u = queue[head]
            head += 1
            
            for v in self.adj[u]:
                if self.color[v] == -1:
                    self.color[v] = 1 - self.color[u]
                    queue.append(v)
                elif self.color[v] == self.color[u]:
                    return False
        return True

if __name__ == "__main__":
    pass

File: without-library/graph/components/bipartite_check.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> adj[200005];
int color[200005];

bool check_bipartite(int start) {
    vector<int> q;
    q.push_back(start);
    int head = 0;
    color[start] = 0;
    
    while(head < q.size()) {
        int u = q[head++];
        
        for(size_t i=0; i<adj[u].size(); i++) {
            int v = adj[u][i];
            if(color[v] == -1) {
                color[v] = 1 - color[u];
                q.push_back(v);
            } else if(color[v] == color[u]) {
                return false;
            }
        }
    }
    return true;
}

int main() {
    return 0;
}
